mail_activity/static/src/js/systray.js

```diff
@@ -0,0 +1,148 @@
+var Widget = require('web.Widget');
+var session = require('web.session');
+var chat_manager = require('mail.chat_manager');
+
+var QWeb = core.qweb;
+
+var Model = require('web.Model');
+
+/**
+ * Menu item appended in the systray part of the navbar, redirects to the next activities of all app
+ */
+var ActivityMenu = Widget.extend({
+    template:'mail.chat.ActivityMenu',
+    events: {
+        "click": "_onActivityMenuClick",
+        "click .o_mail_channel_preview": "_onActivityFilterClick",
+    },
+    start: function () {
+        this.$activities_preview = this.$('.o_mail_navbar_dropdown_channels');
+        chat_manager.bus.on("activity_updated", this, this._updateCounter);
+        chat_manager.is_ready.then(this._updateCounter.bind(this));
+        this._updateActivityPreview();
+        return this._super();
+    },
+
+    // Private
+
+    /**
+     * Make call to model and get current user's activity details
+     * @private
+     */
+    _getActivityData: function(){
+        var self = this;
+        self.model = new Model("res.users")
+        return self.model.call(
+                "activity_user_count",
+                [],
+                {'context': session.user_context}
+                ).then(function(data){
+                    self.activities = data;
+                    self.activityCounter = _.reduce(data, function(total_count, p_data){ return total_count + p_data.total_count; }, 0);
+                    self.$('.o_notification_counter').text(self.activityCounter);
+                    self.$el.toggleClass('o_no_notification', !self.activityCounter);
+                });
+    },
+    /**
+     * Get particular model view to redirect on click of activity scheduled on that model.
+     * @private
+     * @param {string} model
+     */
+    _getActivityModelViewID: function (model) {
+        self.model = new Model(model)
+        return self.model.call("get_activity_view_id", [],{});
+    },
+    /**
+     * Check wether activity systray dropdown is open or not
+     * @private
+     * @returns {boolean}
+     */
+    _isOpen: function () {
+        return this.$el.hasClass('open');
+    },
+    /**
+     * Update(render) activity system tray view on activity updation.
+     * @private
+     */
+    _updateActivityPreview: function () {
+        var self = this;
+        self._getActivityData().then(function (){
+            self.$activities_preview.html(QWeb.render('mail.chat.ActivityMenuPreview', {
+                activities : self.activities
+            }));
+        });
+    },
+    /**
+     * update counter based on activity status(created or Done)
+     * @private
+     * @param {Object} [data] key, value to decide activity created or deleted
+     * @param {String} [data.type] notification type
+     * @param {Boolean} [data.activity_deleted] when activity deleted
+     * @param {Boolean} [data.activity_created] when activity created
+     */
+    _updateCounter: function (data) {
+        if (data) {
```

Comment:
You can also do `if(!data) { return; }` here. Makes the code flatter, but it's nitpicking so not really important :smile:

Reply:
I'd prefer the code as similar to the original as possible